### DISCUSSION TITLE
Support nested dictionary key completion

### DISCUSSION THREAD

**Discussion by javad94** [microsoft/pylance-release#2707] [Enhancement]
## Environment data

-   Language Server version: 2021.6.3
-   OS and version: Windows 10 20H2
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.7.5 using Anaconda

## Expected behaviour

Pylance should show key completion for integer and nested keys.

## Actual behaviour

Pylance does not show key completion for integer and nested keys.

## Logs

```
XXX
```

## Code Snippet / Additional information

```python
somedict = {
    'test': 1,
    'test2':2,
    '3test':3,
    '4':4,
    5: 5,
    '_test6': 6,
    'testnested': {'n1':1, 2:2}
}
```
Pylance does not show key completion for integer keys (in this case 5)
![image](https://user-images.githubusercontent.com/7765309/123192709-b2934900-d4b8-11eb-86d2-da4a2241de05.png)

Pylance does not show key completion for nested keys
![image](https://user-images.githubusercontent.com/7765309/123192762-c939a000-d4b8-11eb-8d0e-6477b84a4389.png)

**Comment by judej** [comment]
Need to investigate why `5` does not show up

**Comment by heejaechang** [comment]
first one is a bug, but nested key is new feature request. so, probably we will fix the first one first,

**Comment by erictraut** [comment]
Our completion provider has logic for string literals but not for other literals like integers, floats, complex, etc.

```python
somedict = {
    5: 0,
    3.41: 0,
    3j: 0,
    [3, 4]: 0,
}
```

We'll need to decide where to draw the line. The vast majority of keys within dicts are strings. It drops off very quickly after that. Integers are probably next.

**Comment by javad94** [comment]
@erictraut I agree with you.

**Comment by cdce8p** [comment]
Aside from strings, I usually see constants that reference strings. These seem to be unsupported at the moment, unfortunately.

![Screen Shot 2021-07-15 at 12 24 35](https://user-images.githubusercontent.com/30130371/125773731-9a856d27-bccc-46b3-a76e-3817b205160f.png)

```py
KEY_1 = "Some key"
KEY_2 = "Some other key"

somedict = {
    KEY_1: 1,
    KEY_2: 2,
    "key_3": 3,
}

somedict[]
```

**Comment by yvvt0379** [comment]
@erictraut I have to explain that in your example, the last key [3, 4] is invalid because it's **mutable**. See the reasons https://docs.python.org/3/faq/design.html#why-must-dictionary-keys-be-immutable

**Comment by erictraut** [comment]
Yes, good point. Replace the list with a tuple.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
